From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: uRyanxD <familiarodrigues123ro@gmail.com>
Date: Sun, 11 Aug 2024 14:43:47 -0300
Subject: [PATCH] Optimise non-flush packet sending


diff --git a/src/main/java/net/minecraft/server/NetworkManager.java b/src/main/java/net/minecraft/server/NetworkManager.java
index da43a6b114ae3af98d4cf98f3a55ce187894c7f9..d69379ce765d5dda640a575b82b1100b58b27b5a 100644
--- a/src/main/java/net/minecraft/server/NetworkManager.java
+++ b/src/main/java/net/minecraft/server/NetworkManager.java
@@ -305,8 +305,7 @@ public class NetworkManager extends SimpleChannelInboundHandler<Packet> {
             }
             // PandaSpigot end
         } else {
-            this.channel.eventLoop().execute(new Runnable() {
-                public void run() {
+            Runnable command = () -> { // PandaSpigot - optimise packets that are not flushed
                     if (enumprotocol != enumprotocol1) {
                         NetworkManager.this.a(enumprotocol);
                     }
@@ -337,8 +336,14 @@ public class NetworkManager extends SimpleChannelInboundHandler<Packet> {
                         packet.onPacketDispatchFinish(player, null);
                     }
                     // PandaSpigot end
-                }
-            });
+            // PandaSpigot start
+            };
+            if (!flush) {
+                io.netty.util.concurrent.AbstractEventExecutor.LazyRunnable run = command::run;
+                this.channel.eventLoop().execute(run);
+            } else {
+                this.channel.eventLoop().execute(command);
+            }
         }
 
     }
