From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Sun, 4 Dec 2022 12:48:52 -0300
Subject: [PATCH] Disable Vanilla Chunk GC

Bukkit has its own system for this.

diff --git a/src/main/java/net/minecraft/server/WorldServer.java b/src/main/java/net/minecraft/server/WorldServer.java
index 02b7b7f87a9b61d00e6b6ba6dd4aa355207ec7b9..841b611bba60f494b362892eea62639ab370aff0 100644
--- a/src/main/java/net/minecraft/server/WorldServer.java
+++ b/src/main/java/net/minecraft/server/WorldServer.java
@@ -962,6 +962,8 @@ public class WorldServer extends World implements IAsyncTaskHandler {
 
             this.chunkProvider.saveChunks(flag, iprogressupdate);
             // CraftBukkit - ArrayList -> Collection
+            // PandaSpigot start - Disable Vanilla Chunk GC
+            /*
             Collection arraylist = this.chunkProviderServer.a();
             Iterator iterator = arraylist.iterator();
 
@@ -972,6 +974,8 @@ public class WorldServer extends World implements IAsyncTaskHandler {
                     this.chunkProviderServer.queueUnload(chunk.locX, chunk.locZ);
                 }
             }
+            */
+            // PandaSpigot end
 
         }
     }
