From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: uRyanxD <familiarodrigues123ro@gmail.com>
Date: Fri, 29 Sep 2023 14:49:41 -0300
Subject: [PATCH] Implement SpongeAbsorbEvent


diff --git a/src/main/java/net/minecraft/server/BlockSponge.java b/src/main/java/net/minecraft/server/BlockSponge.java
index de7714ea58edd5e2e887c50bebd9a5b875676e18..fed87cda8dc181962de578103e01bcb49fffa82c 100644
--- a/src/main/java/net/minecraft/server/BlockSponge.java
+++ b/src/main/java/net/minecraft/server/BlockSponge.java
@@ -46,6 +46,7 @@ public class BlockSponge extends Block {
 
         linkedlist.add(new Tuple(blockposition, Integer.valueOf(0)));
         int i = 0;
+        org.bukkit.craftbukkit.util.BlockStateListPopulator blockList = new org.bukkit.craftbukkit.util.BlockStateListPopulator(world); // PandaSpigot - Use BlockStateListPopulator
 
         BlockPosition blockposition1;
 
@@ -62,7 +63,7 @@ public class BlockSponge extends Block {
                 BlockPosition blockposition2 = blockposition1.shift(enumdirection);
 
                 if (world.getType(blockposition2).getBlock().getMaterial() == Material.WATER) {
-                    world.setTypeAndData(blockposition2, Blocks.AIR.getBlockData(), 2);
+                    blockList.setTypeAndData(blockposition2, Blocks.AIR.getBlockData(), 2); // PandaSpigot
                     arraylist.add(blockposition2);
                     ++i;
                     if (j < 6) {
@@ -83,6 +84,31 @@ public class BlockSponge extends Block {
             world.applyPhysics(blockposition1, Blocks.AIR);
         }
 
+        // PandaSpigot start
+        List<org.bukkit.craftbukkit.block.CraftBlockState> blocks = blockList.getList(); // Is a clone
+        if (!blocks.isEmpty()) {
+            final org.bukkit.Block bblock = world.getWorld().getBlock(blockposition.getX(), blockposition.getY(), blockposition.getZ());
+
+            org.bukkit.event.block.SpongeAbsorbEvent event = new org.bukkit.event.block.SpongeAbsorbEvent(bblock, (List<org.bukkit.block.BlockState>) (List) blocks);
+            if (!event.callEvent()) {
+                return;
+            }
+
+            for (org.bukkit.craftbukkit.block.CraftBlockState block : blocks) {
+                BlockPosition blockposition2 = new BlockPosition(block.getX(), block.getY(), block.getZ());
+                /*
+                IBlockData iblockdata = world.getType(blockposition2);
+                Fluid fluid = world.getFluid(iblockposition2);
+                Material material = iblockdata.getMaterial();
+
+                if (material == Material.WATER) {
+
+                }*/
+                world.setTypeAndData(blockposition2, block.getHandle(), block.getFlag());
+            }
+        }
+        // PandaSpigot end
+
         return i > 0;
     }
 
