From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: uRyanxD <familiarodrigues123ro@gmail.com>
Date: Fri, 15 Nov 2024 15:46:31 -0300
Subject: [PATCH] Remove migration from old versions of the configuration

TODO: Remove spigotconfig migration

diff --git a/src/main/java/org/github/paperspigot/PaperSpigotWorldConfig.java b/src/main/java/org/github/paperspigot/PaperSpigotWorldConfig.java
index 67e56df017b98222e64fc46e97a2b104e6f0d69f..a516990c6fef37b801c918113ebc77082048787c 100644
--- a/src/main/java/org/github/paperspigot/PaperSpigotWorldConfig.java
+++ b/src/main/java/org/github/paperspigot/PaperSpigotWorldConfig.java
@@ -267,6 +267,8 @@ public class PaperSpigotWorldConfig
     public boolean fixCannons;
     private void fixCannons()
     {
+        // PandaSpigot start - Remove migration from old versions of the configuration
+        /*
         // TODO: Remove migrations after most users have upgraded.
         if ( PaperSpigotConfig.version < 9 )
         {
@@ -298,6 +300,8 @@ public class PaperSpigotWorldConfig
                 config.set( "world-settings." + worldName + ".tnt-gameplay", null );
             }
         }
+        */
+        // PandaSpigot end
 
         fixCannons = getBoolean( "fix-cannons", false );
         log( "Fix TNT cannons: " + fixCannons );
